staging/src/k8s.io/kube-aggregator/pkg/util/dynamic_restconfig.go

```diff
@@ -0,0 +1,159 @@
+	egressSelector *egressselector.EgressSelector
+
+	// If present, the Dial method will be used for dialing out to delegate
+	// apiservers.
+	proxyTransport *http.Transport
+
+	insecure   bool
+	serverName string
+	caBundle   []byte
+}
+
+var _ dynamiccertificates.Listener = &DynamicRestConfigProvider{}
+
+// NewDynamicRestConfigProvider creates DynamicRestConfigProvider and subscribes to updates from dynamic certificates
+func NewDynamicRestConfigProvider(servingContent dynamiccertificates.CertKeyContentProvider, egressSelector *egressselector.EgressSelector, proxyTransport *http.Transport, insecure bool, serverName string, caBundle []byte) (*DynamicRestConfigProvider, error) {
+	ret := &DynamicRestConfigProvider{
+		certKeyPair:    servingContent,
+		queue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "DynamicRestConfigProvider"),
+		egressSelector: egressSelector,
+		proxyTransport: proxyTransport,
+		insecure:       insecure,
+		serverName:     serverName,
+		caBundle:       caBundle,
+	}
+
+	if notifier, ok := ret.certKeyPair.(dynamiccertificates.Notifier); ok {
+		notifier.AddListener(ret)
+	}
+
+	// do the initial sync
+	if err := ret.syncRestConfig(); err != nil {
+		return nil, err
+	}
+
+	return ret, nil
+}
+
+// GetRestConfig returns current value of rest config
+func (c *DynamicRestConfigProvider) GetRestConfig() (*restclient.Config, error) {
+	uncastObj := c.currentRestConfig.Load()
+	if uncastObj == nil {
+		return nil, errors.New("dynamicrestconfig: configuration not ready")
+	}
+	restConfig, ok := uncastObj.(*restclient.Config)
+	if !ok {
+		return nil, errors.New("dynamicrestconfig: unexpected config type")
+	}
+
+	return restConfig, nil
+}
+
+// GetTransport returns current value of Transport
+func (c *DynamicRestConfigProvider) GetTransport() (http.RoundTripper, error) {
+	uncastObj := c.currentRestConfig.Load()
+	if uncastObj == nil {
+		return nil, errors.New("dynamicrestconfig: configuration not ready")
+	}
+	transport, ok := uncastObj.(http.RoundTripper)
+	if !ok {
+		return nil, errors.New("dynamicrestconfig: unexpected type for rountripper")
+	}
+
+	return transport, nil
+}
+
+func (c *DynamicRestConfigProvider) syncRestConfig() error {
+	cert, key := c.certKeyPair.CurrentCertKeyContent()
+	restConfig := &restclient.Config{
+		TLSClientConfig: restclient.TLSClientConfig{
+			Insecure:   c.insecure,
+			ServerName: c.serverName,
+			CertData:   cert,
+			KeyData:    key,
+			CAData:     c.caBundle,
+		},
+	}
+	if c.egressSelector != nil {
+		networkContext := egressselector.Cluster.AsNetworkContext()
+		var egressDialer utilnet.DialFunc
+		egressDialer, err := c.egressSelector.Lookup(networkContext)
+		if err != nil {
+			return err
+		}
+		restConfig.Dial = egressDialer
+	} else if c.proxyTransport != nil && c.proxyTransport.DialContext != nil {
+		restConfig.Dial = c.proxyTransport.DialContext
+	}
+	transport, err := restclient.TransportFor(restConfig)
```

Comment:
I think that it will eventually create 'transport.TLSConfig' struct. Could you check if we could use [GetCert](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/transport/config.go#L135) function to dynamically inject the client certificate? Then DynamicRestConfigProvider could provide that function for 'available_controller' and 'available_controller'. WDYT?